File: fc-chat/demo/part3/ch4/message-system-add-push/src/main/java/net/prostars/messagesystem/repository/UserChannelRepository.java
package net.prostars.messagesystem.repository;

import java.util.List;

import net.prostars.messagesystem.dto.projection.ChannelProjection;
import net.prostars.messagesystem.dto.projection.UserIdProjection;
import net.prostars.messagesystem.entity.UserChannelEntity;
import net.prostars.messagesystem.entity.UserChannelId;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;
import org.springframework.lang.NonNull;
import org.springframework.stereotype.Repository;

@Repository
public interface UserChannelRepository extends JpaRepository<UserChannelEntity, UserChannelId> {

  boolean existsByUserIdAndChannelId(@NonNull Long UserId, @NonNull Long ChannelId);

  List<UserIdProjection> findUserIdsByChannelId(@NonNull Long channelId);

  @Query(
      "SELECT c.channelId AS channelId, c.title AS title, c.headCount AS headCount "
          + "FROM UserChannelEntity uc "
          + "INNER JOIN ChannelEntity c ON uc.channelId = c.channelId "
          + "WHERE uc.userId = :userId")
  List<ChannelProjection> findChannelsByUserId(@NonNull @Param("userId") Long userId);
  
  void deleteByUserIdAndChannelId(@NonNull Long userId, @NonNull Long channelId);
}

File: fc-chat/demo/part3/with-push/src/main/kotlin/com/yscorp/withpush/messagesystem/repository/UserChannelRepository.kt
package com.yscorp.withpush.messagesystem.repository

import com.yscorp.withpush.messagesystem.dto.projection.UserIdProjection
import com.yscorp.withpush.messagesystem.entity.UserChannelEntity
import com.yscorp.withpush.messagesystem.entity.UserChannelId
import net.prostars.messagesystem.dto.projection.ChannelProjection
import org.springframework.data.jpa.repository.JpaRepository

@org.springframework.stereotype.Repository
interface UserChannelRepository : JpaRepository<UserChannelEntity, UserChannelId> {
    fun existsByUserIdAndChannelId(
        UserId: Long,
        ChannelId: Long
    ): Boolean

    fun findUserIdsByChannelId(channelId: Long): List<UserIdProjection>

    @org.springframework.data.jpa.repository.Query(
        ("SELECT c.channelId AS channelId, c.title AS title, c.headCount AS headCount "
            + "FROM UserChannelEntity uc "
            + "INNER JOIN ChannelEntity c ON uc.channelId = c.channelId "
            + "WHERE uc.userId = :userId")
    )
    fun findChannelsByUserId(@org.springframework.data.repository.query.Param("userId") userId: Long): List<ChannelProjection>

    fun deleteByUserIdAndChannelId(
        userId: Long,
        channelId: Long
    )
}
